File: utils/plans.ts
import { randomUUID } from 'crypto'
import { copyFile, writeFile } from 'fs/promises'
import memoize from 'lodash-es/memoize.js'
import { join, resolve, sep } from 'path'
import type { AgentId, SessionId } from 'src/types/ids.js'
import type { LogOption } from 'src/types/logs.js'
import type {
  AssistantMessage,
  AttachmentMessage,
  SystemFileSnapshotMessage,
  UserMessage,
} from 'src/types/message.js'
import { getPlanSlugCache, getSessionId } from '../bootstrap/state.js'
import { EXIT_PLAN_MODE_V2_TOOL_NAME } from '../tools/ExitPlanModeTool/constants.js'
import { getCwd } from './cwd.js'
import { logForDebugging } from './debug.js'
import { getMossenConfigHomeDir } from './envUtils.js'
import { isENOENT } from './errors.js'
import { getEnvironmentKind } from './filePersistence/outputsScanner.js'
import { getFsImplementation } from './fsOperations.js'
import { logError } from './log.js'
import { getInitialSettings } from './settings/settings.js'
import { generateWordSlug } from './words.js'
import { validateWorktreeSlug } from './worktree.js'

const MAX_SLUG_RETRIES = 10

// W52 Named Plan Files v0 — prompt-derived ASCII-safe slug.
// Caps below are intentionally smaller than MAX_WORKTREE_SLUG_LENGTH (64) to
// leave room for `-2`/`-3` suffix on collision and the `-agent-{agentId}`
// suffix used by getPlanFilePath for subagents.
const PROMPT_SLUG_MAX_LENGTH = 48
const PROMPT_SLUG_MIN_LENGTH = 2
const PROMPT_SLUG_INPUT_SAMPLE = 120
const PROMPT_SLUG_ASCII_RATIO_FLOOR = 0.5

// ANSI escape sequences. Two forms: CSI (ESC `[ … final-byte`) and OSC
// (ESC `] … BEL|ST`). Stripping both before any text normalization so a
// pasted shell prompt with colors doesn't leak control bytes into the slug.
const ANSI_CSI_PATTERN = /\x1b\[[0-9;?]*[ -/]*[@-~]/g
// eslint-disable-next-line no-control-regex
const ANSI_OSC_PATTERN = /\x1b\][^\x07\x1b]*(?:\x07|\x1b\\)/g

/**
 * Derive an ASCII-safe slug from a user prompt for use as a plan file name.
 *
 * Returns null when the prompt cannot produce a safe slug — caller MUST
 * fall back to {@link generateWordSlug}. Cases that return null:
 *   - empty / whitespace-only input
 *   - mostly non-ASCII content (CJK, emoji-only, etc) — bias check
 *   - result shorter than {@link PROMPT_SLUG_MIN_LENGTH}
 *   - result fails {@link validateWorktreeSlug}
 *
 * Character set is strictly limited to `[a-z0-9-]`. CJK / non-ASCII
 * characters are NOT transliterated; they collapse to the separator. v0
 * design (W52 §2.5 C') keeps slug worktree/branch/file safe at the cost
 * of CJK semantic naming, which a future v2 can add via a separate
 * display-title field decoupled from the slug.
 *
 * Pure: no IO, no network, no session/global state.
 */
export function generatePromptPlanSlug(prompt: string): string | null {
  if (typeof prompt !== 'string') return null

  const trimmed = prompt.trim()
  if (trimmed.length === 0) return null

  // 1. Sample first PROMPT_SLUG_INPUT_SAMPLE chars (after trim) so that
  //    a giant prompt doesn't blow past validateWorktreeSlug's 64-char cap.
  const sample = trimmed.slice(0, PROMPT_SLUG_INPUT_SAMPLE)

  // 2. Strip ANSI escape sequences first — they contain `[`, digits, and
  //    `;` which would otherwise pollute the slug.
  const ansiStripped = sample
    .replace(ANSI_CSI_PATTERN, '')
    .replace(ANSI_OSC_PATTERN, '')

  // 3. Lowercase. Only ASCII letters fold; non-ASCII passes through and is
  //    later collapsed to the separator in step 4.
  const lowered = ansiStripped.toLowerCase()

  // 4. Bias check before normalization: count ASCII alnum in the body and
  //    compare against total non-whitespace chars. If non-ASCII dominates
  //    (CJK paragraph, emoji-only, etc.), return null so caller falls back
  //    to word-slug. This also catches markdown-only / punctuation-only
  //    inputs that would otherwise produce empty slug.
  const asciiAlnumCount = (lowered.match(/[a-z0-9]/g) ?? []).length
  const nonWhitespaceCount = lowered.replace(/\s+/g, '').length
  if (
    nonWhitespaceCount === 0 ||
    asciiAlnumCount / nonWhitespaceCount < PROMPT_SLUG_ASCII_RATIO_FLOOR
  ) {
    return null
  }

  // 5. Collapse anything not in [a-z0-9] to a single dash. This covers
  //    markdown decorations (# * _ ` ~ > | etc), path-dangerous chars
  //    (/ \ : ? * < > | "), CJK, emoji, and all whitespace/punctuation
  //    in one pass.
  const collapsed = lowered.replace(/[^a-z0-9]+/g, '-').replace(/^-+|-+$/g, '')
  if (collapsed.length === 0) return null

  // 6. Truncate to PROMPT_SLUG_MAX_LENGTH and re-strip trailing dashes
  //    introduced by truncation mid-word.
  const truncated = collapsed
    .slice(0, PROMPT_SLUG_MAX_LENGTH)
    .replace(/-+$/, '')
  if (truncated.length < PROMPT_SLUG_MIN_LENGTH) return null

  // 7. Final safety check via worktree slug validator (single source of
  //    truth for worktree/branch/file three-way safety). Rejects anything
  //    that slipped through above.
  try {
    validateWorktreeSlug(truncated)
  } catch {
    return null
  }

  return truncated
}

/**
 * Find a unique slug by trying `base`, then `base-2`, `base-3`, ... up to
 * MAX_SLUG_RETRIES. Returns null if all candidates collide.
 *
 * Caller provides the existence predicate so this function stays pure and
 * testable independent of filesystem state.
 */
function findUniqueSlugWithSuffix(
  base: string,
  exists: (slug: string) => boolean,
): string | null {
  if (!exists(base)) return base
  for (let i = 2; i <= MAX_SLUG_RETRIES; i++) {
    const candidate = `${base}-${i}`
    if (!exists(candidate)) return candidate
  }
  return null
}

/**
 * Get or generate a slug for the current session's plan.
 *
 * The slug is generated lazily on first access and cached for the session.
 *
 * When `options.firstUserPrompt` is supplied AND the prompt yields a safe
 * ASCII slug via {@link generatePromptPlanSlug}, that slug is preferred (with
 * `-2`/`-3` numeric suffix on collision). Otherwise — including the default
 * call shape `getPlanSlug()` / `getPlanSlug(sessionId)` — falls back to the
 * pre-existing word-slug behavior. Collisions in either path retry up to
 * {@link MAX_SLUG_RETRIES} candidates.
 *
 * Backward-compatibility:
 *   - All existing callers (`getPlanSlug()` / `getPlanSlug(id)`) hit the
 *     word-slug branch with byte-for-byte original behavior.
 *   - `setPlanSlug` / `clearPlanSlug` / `copyPlanForResume` /
 *     `copyPlanForFork` are unchanged and never invoke
 *     {@link generatePromptPlanSlug} — resumed/forked sessions reuse the
 *     slug carried by the source log.
 */
export function getPlanSlug(
  sessionId?: SessionId,
  options?: { firstUserPrompt?: string },
): string {
  const id = sessionId ?? getSessionId()
  const cache = getPlanSlugCache()
  let slug = cache.get(id)
  if (!slug) {
    const plansDir = getPlansDirectory()
    const fs = getFsImplementation()
    const exists = (candidate: string): boolean =>
      fs.existsSync(join(plansDir, `${candidate}.md`))

    // Prefer prompt-derived slug when caller passes a usable prompt. Returning
    // null from generatePromptPlanSlug (empty / mostly non-ASCII / fails
    // worktree validator) silently falls through to the word-slug branch
    // below — never throws, never blocks plan creation.
    const promptSeed = options?.firstUserPrompt
    if (typeof promptSeed === 'string' && promptSeed.length > 0) {
      const promptSlug = generatePromptPlanSlug(promptSeed)
      if (promptSlug) {
        const unique = findUniqueSlugWithSuffix(promptSlug, exists)
        if (unique) {
          slug = unique
        }
      }
    }

    // Fallback (also the path for every existing caller): random word slug
    // with collision retry. Behavior identical to the pre-W52 implementation.
    if (!slug) {
      for (let i = 0; i < MAX_SLUG_RETRIES; i++) {
        slug = generateWordSlug()
        const filePath = join(plansDir, `${slug}.md`)
        if (!fs.existsSync(filePath)) {
          break
        }
      }
    }
    cache.set(id, slug!)
  }
  return slug!
}

/**
 * Set a specific plan slug for a session (used when resuming a session)
 */
export function setPlanSlug(sessionId: SessionId, slug: string): void {
  getPlanSlugCache().set(sessionId, slug)
}

/**
 * Clear the plan slug for the current session.
 * This should be called on /clear to ensure a fresh plan file is used.
 */
export function clearPlanSlug(sessionId?: SessionId): void {
  const id = sessionId ?? getSessionId()
  getPlanSlugCache().delete(id)
}

/**
 * Clear ALL plan slug entries (all sessions).
 * Use this on /clear to free sub-session slug entries.
 */
export function clearAllPlanSlugs(): void {
  getPlanSlugCache().clear()
}

// Memoized: called from render bodies (FileReadTool/FileEditTool/FileWriteTool UI.tsx)
// and permission checks. Inputs (initial settings + cwd) are fixed at startup, so the
// mkdirSync result is stable for the session. Without memoization, each rendered tool
// message triggers a mkdirSync syscall (regressed in #20005).
export const getPlansDirectory = memoize(function getPlansDirectory(): string {
  const settings = getInitialSettings()
  const settingsDir = settings.plansDirectory
  let plansPath: string

  if (settingsDir) {
    // Settings.json (relative to project root)
    const cwd = getCwd()
    const resolved = resolve(cwd, settingsDir)

    // Validate path stays within project root to prevent path traversal
    if (!resolved.startsWith(cwd + sep) && resolved !== cwd) {
      logError(
        new Error(`plansDirectory must be within project root: ${settingsDir}`),
      )
      plansPath = join(getMossenConfigHomeDir(), 'plans')
    } else {
      plansPath = resolved
    }
  } else {
    // Default
    plansPath = join(getMossenConfigHomeDir(), 'plans')
  }

  // Ensure directory exists (mkdirSync with recursive: true is a no-op if it exists)
  try {
    getFsImplementation().mkdirSync(plansPath)
  } catch (error) {
    logError(error)
  }

  return plansPath
})

/**
 * Get the file path for a session's plan
 * @param agentId Optional agent ID for subagents. If not provided, returns main session plan.
 * For main conversation (no agentId), returns {planSlug}.md
 * For subagents (agentId provided), returns {planSlug}-agent-{agentId}.md
 */
export function getPlanFilePath(agentId?: AgentId): string {
  const planSlug = getPlanSlug(getSessionId())

  // Main conversation: simple filename with word slug
  if (!agentId) {
    return join(getPlansDirectory(), `${planSlug}.md`)
  }

  // Subagents: include agent ID
  return join(getPlansDirectory(), `${planSlug}-agent-${agentId}.md`)
}

/**
 * Get the plan content for a session
 * @param agentId Optional agent ID for subagents. If not provided, returns main session plan.
 */
export function getPlan(agentId?: AgentId): string | null {
  const filePath = getPlanFilePath(agentId)
  try {
    return getFsImplementation().readFileSync(filePath, { encoding: 'utf-8' })
  } catch (error) {
    if (isENOENT(error)) return null
    logError(error)
    return null
  }
}

/**
 * Extract the plan slug from a log's message history.
 */
function getSlugFromLog(log: LogOption): string | undefined {
  return log.messages.find(m => m.slug)?.slug
}

/**
 * Restore plan slug from a resumed session.
 * Sets the slug in the session cache so getPlanSlug returns it.
 * If the plan file is missing, attempts to recover it from a file snapshot
 * (written incrementally during the session) or from message history.
 * Returns true if a plan file exists (or was recovered) for the slug.
 * @param log The log to restore from
 * @param targetSessionId The session ID to associate the plan slug with.
 *                        This should be the ORIGINAL session ID being resumed,
 *                        not the temporary session ID from before resume.
 */
export async function copyPlanForResume(
  log: LogOption,
  targetSessionId?: SessionId,
): Promise<boolean> {
  const slug = getSlugFromLog(log)
  if (!slug) {
    return false
  }

  // Set the slug for the target session ID (or current if not provided)
  const sessionId = targetSessionId ?? getSessionId()
  setPlanSlug(sessionId, slug)

  // Attempt to read the plan file directly — recovery triggers on ENOENT.
  const planPath = join(getPlansDirectory(), `${slug}.md`)
  try {
    await getFsImplementation().readFile(planPath, { encoding: 'utf-8' })
    return true
  } catch (e: unknown) {
    if (!isENOENT(e)) {
      // Don't throw — called fire-and-forget (void copyPlanForResume(...)) with no .catch()
      logError(e)
      return false
    }
    // Only attempt recovery in remote sessions (CCR) where files don't persist
    if (getEnvironmentKind() === null) {
      return false
    }

    logForDebugging(
      `Plan file missing during resume: ${planPath}. Attempting recovery.`,
    )

    // Try file snapshot first (written incrementally during session)
    const snapshotPlan = findFileSnapshotEntry(log.messages, 'plan')
    let recovered: string | null = null
    if (snapshotPlan && snapshotPlan.content.length > 0) {
      recovered = snapshotPlan.content
      logForDebugging(
        `Plan recovered from file snapshot, ${recovered.length} chars`,
        { level: 'info' },
      )
    } else {
      // Fall back to searching message history
      recovered = recoverPlanFromMessages(log)
      if (recovered) {
        logForDebugging(
          `Plan recovered from message history, ${recovered.length} chars`,
          { level: 'info' },
        )
      }
    }

    if (recovered) {
      try {
        await writeFile(planPath, recovered, { encoding: 'utf-8' })
        return true
      } catch (writeError) {
        logError(writeError)
        return false
      }
    }
    logForDebugging(
      'Plan file recovery failed: no file snapshot or plan content found in message history',
    )
    return false
  }
}

/**
 * Copy a plan file for a forked session. Unlike copyPlanForResume (which reuses
 * the original slug), this generates a NEW slug for the forked session and
 * writes the original plan content to the new file. This prevents the original
 * and forked sessions from clobbering each other's plan files.
 */
export async function copyPlanForFork(
  log: LogOption,
  targetSessionId: SessionId,
): Promise<boolean> {
  const originalSlug = getSlugFromLog(log)
  if (!originalSlug) {
    return false
  }

  const plansDir = getPlansDirectory()
  const originalPlanPath = join(plansDir, `${originalSlug}.md`)

  // Generate a new slug for the forked session (do NOT reuse the original)
  const newSlug = getPlanSlug(targetSessionId)
  const newPlanPath = join(plansDir, `${newSlug}.md`)
  try {
    await copyFile(originalPlanPath, newPlanPath)
    return true
  } catch (error) {
    if (isENOENT(error)) {
      return false
    }
    logError(error)
    return false
  }
}

/**
 * Recover plan content from the message history. Plan content can appear in
 * three forms depending on what happened during the session:
 *
 * 1. ExitPlanMode tool_use input — normalizeToolInput injects the plan content
 *    into the tool_use input, which persists in the transcript.
 *
 * 2. planContent field on user messages — set during the "clear context and
 *    implement" flow when ExitPlanMode is approved.
 *
 * 3. plan_file_reference attachment — created by auto-compact to preserve the
 *    plan across compaction boundaries.
 */
function recoverPlanFromMessages(log: LogOption): string | null {
  for (let i = log.messages.length - 1; i >= 0; i--) {
    const msg = log.messages[i]
    if (!msg) {
      continue
    }

    if (msg.type === 'assistant') {
      const { content } = (msg as AssistantMessage).message
      if (Array.isArray(content)) {
        for (const block of content) {
          if (
            block.type === 'tool_use' &&
            block.name === EXIT_PLAN_MODE_V2_TOOL_NAME
          ) {
            const input = block.input as Record<string, unknown> | undefined
            const plan = input?.plan
            if (typeof plan === 'string' && plan.length > 0) {
              return plan
            }
          }
        }
      }
    }

    if (msg.type === 'user') {
      const userMsg = msg as UserMessage
      if (
        typeof userMsg.planContent === 'string' &&
        userMsg.planContent.length > 0
      ) {
        return userMsg.planContent
      }
    }

    if (msg.type === 'attachment') {
      const attachmentMsg = msg as AttachmentMessage
      if (attachmentMsg.attachment?.type === 'plan_file_reference') {
        const plan = (attachmentMsg.attachment as { planContent?: string })
          .planContent
        if (typeof plan === 'string' && plan.length > 0) {
          return plan
        }
      }
    }
  }
  return null
}

/**
 * Find a file entry in the most recent file-snapshot system message in the transcript.
 * Scans backwards to find the latest snapshot.
 */
function findFileSnapshotEntry(
  messages: LogOption['messages'],
  key: string,
): { key: string; path: string; content: string } | undefined {
  for (let i = messages.length - 1; i >= 0; i--) {
    const msg = messages[i]
    if (
      msg?.type === 'system' &&
      'subtype' in msg &&
      msg.subtype === 'file_snapshot' &&
      'snapshotFiles' in msg
    ) {
      const files = msg.snapshotFiles as Array<{
        key: string
        path: string
        content: string
      }>
      return files.find(f => f.key === key)
    }
  }
  return undefined
}

/**
 * Persist a snapshot of session files (plan, todos) to the transcript.
 * Called incrementally whenever these files change. Only active in remote
 * sessions (CCR) where local files don't persist between sessions.
 */
export async function persistFileSnapshotIfRemote(): Promise<void> {
  if (getEnvironmentKind() === null) {
    return
  }
  try {
    const snapshotFiles: SystemFileSnapshotMessage['snapshotFiles'] = []

    // Snapshot plan file
    const plan = getPlan()
    if (plan) {
      snapshotFiles.push({
        key: 'plan',
        path: getPlanFilePath(),
        content: plan,
      })
    }

    if (snapshotFiles.length === 0) {
      return
    }

    const message: SystemFileSnapshotMessage = {
      type: 'system',
      subtype: 'file_snapshot',
      content: 'File snapshot',
      level: 'info',
      isMeta: true,
      timestamp: new Date().toISOString(),
      uuid: randomUUID(),
      snapshotFiles,
    }

    const { recordTranscript } = await import('./sessionStorage.js')
    await recordTranscript([message])
  } catch (error) {
    logError(error)
  }
}

File: crates/mossen-utils/src/plans.rs
use regex::Regex;
use std::collections::HashMap;
use std::path::{Path, PathBuf};

const MAX_SLUG_RETRIES: usize = 10;
const PROMPT_SLUG_MAX_LENGTH: usize = 48;
const PROMPT_SLUG_MIN_LENGTH: usize = 2;
const PROMPT_SLUG_INPUT_SAMPLE: usize = 120;
const PROMPT_SLUG_ASCII_RATIO_FLOOR: f64 = 0.5;

/// Derive an ASCII-safe slug from a user prompt for use as a plan file name.
/// Returns None when the prompt cannot produce a safe slug.
pub fn generate_prompt_plan_slug(prompt: &str) -> Option<String> {
    let trimmed = prompt.trim();
    if trimmed.is_empty() {
        return None;
    }

    // 1. Sample first PROMPT_SLUG_INPUT_SAMPLE chars
    let sample: String = trimmed.chars().take(PROMPT_SLUG_INPUT_SAMPLE).collect();

    // 2. Strip ANSI escape sequences
    let ansi_csi_re = Regex::new(r"\x1b\[[0-9;?]*[ -/]*[@-~]").unwrap();
    let ansi_osc_re = Regex::new(r"\x1b\][^\x07\x1b]*(?:\x07|\x1b\\)").unwrap();
    let ansi_stripped = ansi_csi_re.replace_all(&sample, "");
    let ansi_stripped = ansi_osc_re.replace_all(&ansi_stripped, "");

    // 3. Lowercase
    let lowered = ansi_stripped.to_lowercase();

    // 4. Bias check: count ASCII alnum
    let ascii_alnum_count = lowered
        .chars()
        .filter(|c| c.is_ascii_alphanumeric())
        .count();
    let non_whitespace_count = lowered.chars().filter(|c| !c.is_whitespace()).count();
    if non_whitespace_count == 0
        || (ascii_alnum_count as f64 / non_whitespace_count as f64) < PROMPT_SLUG_ASCII_RATIO_FLOOR
    {
        return None;
    }

    // 5. Collapse anything not in [a-z0-9] to a single dash
    let non_alnum_re = Regex::new(r"[^a-z0-9]+").unwrap();
    let collapsed = non_alnum_re.replace_all(&lowered, "-");
    let collapsed = collapsed.trim_matches('-');
    if collapsed.is_empty() {
        return None;
    }

    // 6. Truncate to PROMPT_SLUG_MAX_LENGTH
    let truncated: String = collapsed.chars().take(PROMPT_SLUG_MAX_LENGTH).collect();
    let truncated = truncated.trim_end_matches('-');
    if truncated.len() < PROMPT_SLUG_MIN_LENGTH {
        return None;
    }

    // 7. Final safety check via worktree slug validator
    if !validate_worktree_slug(truncated) {
        return None;
    }

    Some(truncated.to_string())
}

/// Validate that a slug is safe for worktree/branch/file use.
fn validate_worktree_slug(slug: &str) -> bool {
    if slug.is_empty() || slug.len() > 64 {
        return false;
    }
    // Must only contain [a-z0-9-]
    let valid_re = Regex::new(r"^[a-z0-9-]+$").unwrap();
    if !valid_re.is_match(slug) {
        return false;
    }
    // Must not start or end with dash
    if slug.starts_with('-') || slug.ends_with('-') {
        return false;
    }
    // Must not contain consecutive dashes
    if slug.contains("--") {
        return false;
    }
    true
}

/// Find a unique slug by trying `base`, then `base-2`, `base-3`, ...
fn find_unique_slug_with_suffix<F>(base: &str, exists: F) -> Option<String>
where
    F: Fn(&str) -> bool,
{
    if !exists(base) {
        return Some(base.to_string());
    }
    for i in 2..=MAX_SLUG_RETRIES {
        let candidate = format!("{}-{}", base, i);
        if !exists(&candidate) {
            return Some(candidate);
        }
    }
    None
}

/// Plan slug cache, mapping session IDs to slugs.
pub struct PlanSlugCache {
    cache: HashMap<String, String>,
}

impl PlanSlugCache {
    pub fn new() -> Self {
        Self {
            cache: HashMap::new(),
        }
    }

    pub fn get(&self, session_id: &str) -> Option<&String> {
        self.cache.get(session_id)
    }

    pub fn set(&mut self, session_id: String, slug: String) {
        self.cache.insert(session_id, slug);
    }

    pub fn delete(&mut self, session_id: &str) {
        self.cache.remove(session_id);
    }

    pub fn clear(&mut self) {
        self.cache.clear();
    }
}

impl Default for PlanSlugCache {
    fn default() -> Self {
        Self::new()
    }
}

/// Get or generate a slug for the current session's plan.
pub fn get_plan_slug(
    session_id: &str,
    cache: &mut PlanSlugCache,
    plans_dir: &Path,
    first_user_prompt: Option<&str>,
) -> String {
    if let Some(slug) = cache.get(session_id) {
        return slug.clone();
    }

    let exists = |candidate: &str| -> bool { plans_dir.join(format!("{}.md", candidate)).exists() };

    let mut slug: Option<String> = None;

    // Prefer prompt-derived slug when caller passes a usable prompt
    if let Some(prompt_seed) = first_user_prompt {
        if !prompt_seed.is_empty() {
            if let Some(prompt_slug) = generate_prompt_plan_slug(prompt_seed) {
                if let Some(unique) = find_unique_slug_with_suffix(&prompt_slug, exists) {
                    slug = Some(unique);
                }
            }
        }
    }

    // Fallback: random word slug with collision retry
    if slug.is_none() {
        for _ in 0..MAX_SLUG_RETRIES {
            let candidate = generate_word_slug();
            let file_path = plans_dir.join(format!("{}.md", candidate));
            if !file_path.exists() {
                slug = Some(candidate);
                break;
            }
        }
    }

    let slug = slug.unwrap_or_else(generate_word_slug);
    cache.set(session_id.to_string(), slug.clone());
    slug
}

/// Set a specific plan slug for a session (used when resuming a session).
pub fn set_plan_slug(cache: &mut PlanSlugCache, session_id: &str, slug: &str) {
    cache.set(session_id.to_string(), slug.to_string());
}

/// Clear the plan slug for a session.
pub fn clear_plan_slug(cache: &mut PlanSlugCache, session_id: &str) {
    cache.delete(session_id);
}

/// Clear ALL plan slug entries (all sessions).
pub fn clear_all_plan_slugs(cache: &mut PlanSlugCache) {
    cache.clear();
}

/// Get the plans directory path.
pub fn get_plans_directory(
    settings_plans_dir: Option<&str>,
    cwd: &Path,
    config_home_dir: &Path,
) -> PathBuf {
    if let Some(settings_dir) = settings_plans_dir {
        let resolved = cwd.join(settings_dir);
        let cwd_str = cwd.to_string_lossy();
        let resolved_str = resolved.to_string_lossy();

        // Validate path stays within project root
        if !resolved_str.starts_with(cwd_str.as_ref()) && resolved != *cwd {
            eprintln!(
                "plansDirectory must be within project root: {}",
                settings_dir
            );
            let plans_path = config_home_dir.join("plans");
            let _ = std::fs::create_dir_all(&plans_path);
            return plans_path;
        }

        let _ = std::fs::create_dir_all(&resolved);
        resolved
    } else {
        let plans_path = config_home_dir.join("plans");
        let _ = std::fs::create_dir_all(&plans_path);
        plans_path
    }
}

/// Get the file path for a session's plan.
pub fn get_plan_file_path(plan_slug: &str, plans_dir: &Path, agent_id: Option<&str>) -> PathBuf {
    match agent_id {
        None => plans_dir.join(format!("{}.md", plan_slug)),
        Some(id) => plans_dir.join(format!("{}-agent-{}.md", plan_slug, id)),
    }
}

/// Get the plan content for a session.
pub fn get_plan(plan_file_path: &Path) -> Option<String> {
    match std::fs::read_to_string(plan_file_path) {
        Ok(content) => Some(content),
        Err(e) if e.kind() == std::io::ErrorKind::NotFound => None,
        Err(e) => {
            eprintln!("Error reading plan file: {}", e);
            None
        }
    }
}

/// Extract the plan slug from a log's message history.
pub fn get_slug_from_log(messages: &[LogMessage]) -> Option<String> {
    messages.iter().find_map(|m| m.slug.clone())
}

/// A log message with optional slug.
#[derive(Debug, Clone)]
pub struct LogMessage {
    pub slug: Option<String>,
    pub message_type: String,
    pub content: serde_json::Value,
}

/// Log option for plan operations.
#[derive(Debug, Clone)]
pub struct LogOption {
    pub messages: Vec<LogMessage>,
}

/// Copy a plan file for resume. Sets the slug in the session cache.
/// Returns true if a plan file exists (or was recovered) for the slug.
pub async fn copy_plan_for_resume(
    log: &LogOption,
    target_session_id: &str,
    cache: &mut PlanSlugCache,
    plans_dir: &Path,
) -> bool {
    let slug = match get_slug_from_log(&log.messages) {
        Some(s) => s,
        None => return false,
    };

    set_plan_slug(cache, target_session_id, &slug);

    let plan_path = plans_dir.join(format!("{}.md", slug));
    match tokio::fs::read_to_string(&plan_path).await {
        Ok(_) => true,
        Err(e) if e.kind() == std::io::ErrorKind::NotFound => {
            // Try recovery from message history
            if let Some(recovered) = recover_plan_from_messages(log) {
                match tokio::fs::write(&plan_path, &recovered).await {
                    Ok(_) => true,
                    Err(_) => false,
                }
            } else {
                false
            }
        }
        Err(_) => false,
    }
}

/// Copy a plan file for a forked session.
pub async fn copy_plan_for_fork(
    log: &LogOption,
    target_session_id: &str,
    cache: &mut PlanSlugCache,
    plans_dir: &Path,
) -> bool {
    let original_slug = match get_slug_from_log(&log.messages) {
        Some(s) => s,
        None => return false,
    };

    let original_plan_path = plans_dir.join(format!("{}.md", original_slug));
    let new_slug = get_plan_slug(target_session_id, cache, plans_dir, None);
    let new_plan_path = plans_dir.join(format!("{}.md", new_slug));

    match tokio::fs::copy(&original_plan_path, &new_plan_path).await {
        Ok(_) => true,
        Err(e) if e.kind() == std::io::ErrorKind::NotFound => false,
        Err(e) => {
            eprintln!("Error copying plan file: {}", e);
            false
        }
    }
}

/// Recover plan content from the message history.
fn recover_plan_from_messages(log: &LogOption) -> Option<String> {
    for msg in log.messages.iter().rev() {
        if msg.message_type == "assistant" {
            if let Some(content) = msg.content.as_array() {
                for block in content {
                    if block.get("type").and_then(|v| v.as_str()) == Some("tool_use")
                        && block.get("name").and_then(|v| v.as_str()) == Some("ExitPlanMode_v2")
                    {
                        if let Some(input) = block.get("input") {
                            if let Some(plan) = input.get("plan").and_then(|v| v.as_str()) {
                                if !plan.is_empty() {
                                    return Some(plan.to_string());
                                }
                            }
                        }
                    }
                }
            }
        }

        if msg.message_type == "user" {
            if let Some(plan_content) = msg.content.get("planContent").and_then(|v| v.as_str()) {
                if !plan_content.is_empty() {
                    return Some(plan_content.to_string());
                }
            }
        }

        if msg.message_type == "attachment" {
            if let Some(attachment) = msg.content.get("attachment") {
                if attachment.get("type").and_then(|v| v.as_str()) == Some("plan_file_reference") {
                    if let Some(plan) = attachment.get("planContent").and_then(|v| v.as_str()) {
                        if !plan.is_empty() {
                            return Some(plan.to_string());
                        }
                    }
                }
            }
        }
    }
    None
}

/// Find a file entry in the most recent file-snapshot system message.
pub fn find_file_snapshot_entry(messages: &[LogMessage], key: &str) -> Option<FileSnapshotEntry> {
    for msg in messages.iter().rev() {
        if msg.message_type == "system" {
            if let Some(subtype) = msg.content.get("subtype").and_then(|v| v.as_str()) {
                if subtype == "file_snapshot" {
                    if let Some(files) = msg.content.get("snapshotFiles").and_then(|v| v.as_array())
                    {
                        for file in files {
                            if file.get("key").and_then(|v| v.as_str()) == Some(key) {
                                return Some(FileSnapshotEntry {
                                    key: key.to_string(),
                                    path: file
                                        .get("path")
                                        .and_then(|v| v.as_str())
                                        .unwrap_or("")
                                        .to_string(),
                                    content: file
                                        .get("content")
                                        .and_then(|v| v.as_str())
                                        .unwrap_or("")
                                        .to_string(),
                                });
                            }
                        }
                    }
                }
            }
        }
    }
    None
}

/// A file snapshot entry.
#[derive(Debug, Clone)]
pub struct FileSnapshotEntry {
    pub key: String,
    pub path: String,
    pub content: String,
}

/// Persist a snapshot of session files to the transcript (only in remote sessions).
pub async fn persist_file_snapshot_if_remote(
    plan: Option<&str>,
    plan_file_path: Option<&Path>,
    is_remote: bool,
) -> Result<(), Box<dyn std::error::Error>> {
    if !is_remote {
        return Ok(());
    }

    let mut snapshot_files = Vec::new();

    if let (Some(plan_content), Some(path)) = (plan, plan_file_path) {
        snapshot_files.push(serde_json::json!({
            "key": "plan",
            "path": path.to_string_lossy(),
            "content": plan_content,
        }));
    }

    if snapshot_files.is_empty() {
        return Ok(());
    }

    // Record the snapshot message
    // In a real implementation, this would write to the session transcript
    Ok(())
}

/// Generate a random word slug (placeholder implementation).
fn generate_word_slug() -> String {
    use std::time::{SystemTime, UNIX_EPOCH};
    let seed = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .subsec_nanos();
    format!("plan-{:08x}", seed)
}
